File: ciris_sdk/resources/system.py
from __future__ import annotations

from typing import List, Optional, Dict, Any, Union

from ..models import (
    TelemetrySnapshot, SystemHealth, ServiceInfo, ProcessorState,
    MetricRecord, AdapterInfo
)
from ..transport import Transport

class SystemResource:
    def __init__(self, transport: Transport):
        self._transport = transport

    # Telemetry Methods
    async def get_telemetry_snapshot(self) -> TelemetrySnapshot:
        """Get complete system telemetry snapshot."""
        resp = await self._transport.request("GET", "/v1/system/telemetry")
        return TelemetrySnapshot(**resp.json())

    async def get_adapters_info(self) -> List[AdapterInfo]:
        """Get information about all registered adapters."""
        resp = await self._transport.request("GET", "/v1/system/adapters")
        adapters = []
        for adapter_data in resp.json():
            adapters.append(AdapterInfo(**adapter_data))
        return adapters

    async def get_services_info(self) -> List[ServiceInfo]:
        """Get information about all registered services."""
        resp = await self._transport.request("GET", "/v1/system/services")
        services = []
        for service_data in resp.json():
            services.append(ServiceInfo(**service_data))
        return services

    async def get_processor_state(self) -> ProcessorState:
        """Get current processor state information."""
        resp = await self._transport.request("GET", "/v1/system/processor/state")
        return ProcessorState(**resp.json())

    async def get_configuration_snapshot(self) -> Dict[str, Any]:
        """Get current system configuration snapshot."""
        resp = await self._transport.request("GET", "/v1/system/configuration")
        return resp.json()

    async def get_health_status(self) -> SystemHealth:
        """Get overall system health status."""
        resp = await self._transport.request("GET", "/v1/system/health")
        return SystemHealth(**resp.json())

    # Processor Control Methods (System endpoints)
    async def single_step(self) -> Dict[str, Any]:
        """Execute a single processing step via system endpoint."""
        resp = await self._transport.request("POST", "/v1/system/processor/step")
        return resp.json()

    async def pause_processing(self) -> Dict[str, Any]:
        """Pause the processor via system endpoint."""
        resp = await self._transport.request("POST", "/v1/system/processor/pause")
        return resp.json()

    async def resume_processing(self) -> Dict[str, Any]:
        """Resume the processor via system endpoint."""
        resp = await self._transport.request("POST", "/v1/system/processor/resume")
        return resp.json()

    async def get_processing_queue_status(self) -> Dict[str, Any]:
        """Get current processing queue status via system endpoint."""
        resp = await self._transport.request("GET", "/v1/system/processor/queue")
        return resp.json()

    # Metrics Methods
    async def record_metric(
        self,
        metric_name: str,
        value: Union[int, float],
        tags: Optional[Dict[str, str]] = None
    ) -> Dict[str, Any]:
        """Record a custom metric with optional tags."""
        payload = {
            "metric_name": metric_name,
            "value": value,
            "tags": tags or {}
        }
        resp = await self._transport.request("POST", "/v1/system/metrics", json=payload)
        return resp.json()

    async def get_metrics_history(
        self,
        metric_name: str,
        hours: int = 24
    ) -> List[Dict[str, Any]]:
        """Get historical data for a specific metric."""
        resp = await self._transport.request(
            "GET",
            f"/v1/system/metrics/{metric_name}/history",
            params={"hours": str(hours)}
        )
        return resp.json()

File: CIRISGUI/apps/agui/lib/ciris-sdk/resources/system.ts
// CIRIS TypeScript SDK - System Resource

import { BaseResource } from './base';
import {
  HealthStatus,
  ServiceInfo,
  ResourceUsage,
  ProcessorQueueStatus,
  RuntimeControlExtendedResponse,
  ServiceHealthStatus,
  ServicePriorityUpdateRequest,
  CircuitBreakerResetRequest,
  ServiceSelectionExplanation,
  ProcessorStateInfo
} from '../types';

export interface RuntimeControlResponse {
  status: string;
  message: string;
  timestamp: string;
}


export interface AdapterInfo {
  adapter_id: string;
  adapter_type: string;
  is_running: boolean;
  channels: string[];
  message_count: number;
  error_count: number;
  created_at: string;
  last_activity?: string;
  config?: Record<string, any>;
}

export interface AdapterListResponse {
  adapters: AdapterInfo[];
  total_count: number;
  running_count: number;
}

export interface AdapterOperationResult {
  success: boolean;
  adapter_id?: string;
  message: string;
  adapter_type?: string;
}

export interface RegisterAdapterRequest {
  config?: Record<string, any>;
}

export class SystemResource extends BaseResource {
  /**
   * Get system health status
   */
  async getHealth(): Promise<HealthStatus> {
    return this.transport.get<HealthStatus>('/v1/system/health');
  }

  /**
   * Get all services status
   */
  async getServices(): Promise<{
    services: ServiceInfo[];
    total_services: number;
    healthy_services: number;
    timestamp: string;
  }> {
    return this.transport.get('/v1/system/services');
  }

  /**
   * Get resource usage
   */
  async getResources(): Promise<ResourceUsage> {
    return this.transport.get<ResourceUsage>('/v1/system/resources');
  }

  /**
   * Pause runtime processing
   */
  async pauseRuntime(): Promise<RuntimeControlResponse> {
    return this.transport.post<RuntimeControlResponse>('/v1/system/runtime/pause');
  }

  /**
   * Resume runtime processing
   */
  async resumeRuntime(): Promise<RuntimeControlResponse> {
    return this.transport.post<RuntimeControlResponse>('/v1/system/runtime/resume');
  }

  /**
   * Get runtime status
   */
  async getRuntimeStatus(): Promise<{
    is_paused: boolean;
    pause_reason?: string;
    paused_at?: string;
    paused_by?: string;
    processor_status?: string;
    health_status?: string;
    uptime_seconds?: number;
    active_adapters?: any[];
    loaded_adapters?: any[];
  }> {
    // Use the runtime state endpoint to get status
    const response = await this.transport.post('/v1/system/runtime/state', {});
    return {
      is_paused: response.processor_state === 'paused',
      processor_status: response.processor_state,
      health_status: 'healthy', // Not available in state endpoint
      uptime_seconds: 0, // Not available in state endpoint
      active_adapters: [],
      loaded_adapters: []
    };
  }

  /**
   * Get runtime state
   */
  async getRuntimeState(): Promise<{
    success: boolean;
    message: string;
    processor_state: string;
    cognitive_state: string;
    queue_depth: number;
  }> {
    return this.transport.post('/v1/system/runtime/state', {});
  }


  /**
   * Pause a specific processor
   */
  async pauseProcessor(processorName: string, duration?: number): Promise<RuntimeControlResponse> {
    return this.transport.post<RuntimeControlResponse>(
      `/v1/system/processors/${processorName}/pause`,
      { duration }
    );
  }

  /**
   * Resume a specific processor
   */
  async resumeProcessor(processorName: string): Promise<RuntimeControlResponse> {
    return this.transport.post<RuntimeControlResponse>(
      `/v1/system/processors/${processorName}/resume`
    );
  }

  /**
   * Get all adapters
   */
  async getAdapters(): Promise<AdapterListResponse> {
    return this.transport.get<AdapterListResponse>('/v1/system/adapters');
  }

  /**
   * Get a specific adapter
   */
  async getAdapter(adapterId: string): Promise<AdapterInfo> {
    return this.transport.get<AdapterInfo>(`/v1/system/adapters/${adapterId}`);
  }

  /**
   * Register a new adapter
   */
  async registerAdapter(
    adapterType: string,
    config?: Record<string, any>
  ): Promise<AdapterOperationResult> {
    return this.transport.post<AdapterOperationResult>(
      `/v1/system/adapters/${adapterType}`,
      { config }
    );
  }

  /**
   * Unregister an adapter
   */
  async unregisterAdapter(adapterId: string): Promise<AdapterOperationResult> {
    return this.transport.delete<AdapterOperationResult>(
      `/v1/system/adapters/${adapterId}`
    );
  }

  /**
   * Reload an adapter
   */
  async reloadAdapter(adapterId: string): Promise<AdapterOperationResult> {
    return this.transport.put<AdapterOperationResult>(
      `/v1/system/adapters/${adapterId}/reload`
    );
  }

  /**
   * Restart a service
   */
  async restartService(serviceName: string): Promise<RuntimeControlResponse> {
    return this.transport.post<RuntimeControlResponse>(
      `/v1/system/services/${serviceName}/restart`
    );
  }

  /**
   * Pause an adapter
   */
  async pauseAdapter(adapterName: string, duration?: number): Promise<RuntimeControlResponse> {
    return this.transport.post<RuntimeControlResponse>(
      `/v1/system/adapters/${adapterName}/pause`,
      { duration }
    );
  }

  /**
   * Resume an adapter
   */
  async resumeAdapter(adapterName: string): Promise<RuntimeControlResponse> {
    return this.transport.post<RuntimeControlResponse>(
      `/v1/system/adapters/${adapterName}/resume`
    );
  }

  // Extended System Management Methods

  /**
   * Get processing queue status
   */
  async getProcessingQueueStatus(): Promise<ProcessorQueueStatus> {
    return this.transport.get<ProcessorQueueStatus>('/v1/system/runtime/queue');
  }

  /**
   * Execute a single processing step
   */
  async singleStepProcessor(): Promise<RuntimeControlExtendedResponse> {
    return this.transport.post<RuntimeControlExtendedResponse>('/v1/system/runtime/step');
  }

  /**
   * Get detailed service health status
   */
  async getServiceHealthDetails(): Promise<ServiceHealthStatus> {
    return this.transport.get<ServiceHealthStatus>('/v1/system/services/health');
  }

  /**
   * Update service provider priority
   */
  async updateServicePriority(
    providerName: string,
    update: ServicePriorityUpdateRequest
  ): Promise<{ success: boolean; message?: string; error?: string }> {
    return this.transport.put(
      `/v1/system/services/${providerName}/priority`,
      update
    );
  }

  /**
   * Reset circuit breakers
   */
  async resetCircuitBreakers(
    request: CircuitBreakerResetRequest = {}
  ): Promise<{ success: boolean; message?: string; reset_count?: number }> {
    return this.transport.post('/v1/system/services/circuit-breakers/reset', request);
  }

  /**
   * Get service selection logic explanation
   */
  async getServiceSelectionExplanation(): Promise<ServiceSelectionExplanation> {
    return this.transport.get<ServiceSelectionExplanation>('/v1/system/services/selection-logic');
  }

  /**
   * Get information about all processor states
   */
  async getProcessorStates(): Promise<ProcessorStateInfo[]> {
    return this.transport.get<ProcessorStateInfo[]>('/v1/system/processors');
  }

  /**
   * Get current system time
   */
  async getTime(): Promise<{
    current_time: string;
    timezone: string;
    timestamp: number;
  }> {
    return this.transport.get('/v1/system/time');
  }

  /**
   * Initiate system shutdown
   */
  async shutdown(reason?: string): Promise<{
    status: string;
    message: string;
    shutdown_initiated: boolean;
  }> {
    return this.transport.post('/v1/system/shutdown', { reason });
  }
}